packages/core/graph/src/AdjacencyList.js

```diff
@@ -0,0 +1,1210 @@
+ * The buffer is partitioned into:
+ * - a header, which stores the capacity and number of items in the map,
+ * - a hash table, which is an array of pointers to linked lists of items
+ *   with the same hash,
+ * - an items array, which is where the linked items are stored.
+ *
+ *            hash table                 item
+ *            (capacity)             (ITEM_SIZE)
+ *         ┌──────┴──────┐             ┌──┴──┐
+ *   ┌──┬──┬──┬───────┬──┬──┬──┬───────┬──┬──┐
+ *   │  │  │  │  ...  │  │  │  │  ...  │  │  │
+ *   └──┴──┴──┴───────┴──┴──┴──┴───────┴──┴──┘
+ *   └──┬──┘             └─────────┬─────────┘
+ *    header                     items
+ * (HEADER_SIZE)    (capacity * ITEM_SIZE * BUCKET_SIZE)
+ *
+ *
+ * An item is added with a hash key that fits within the range of the hash
+ * table capacity. The item is stored at the next available address after the
+ * hash table, and a pointer to the address is stored in the hash table at
+ * the index matching the hash. If the hash is already pointing at an item,
+ * the pointer is stored in the `next` field of the existing item instead.
+ *
+ *       hash table                          items
+ * ┌─────────┴────────┐┌───────────────────────┴────────────────────────┐
+ *    0    1    2        11       17        23       29      35
+ * ┌───┐┌───┐┌───┐┌───┐┌───┬───┐┌───┬───┐┌───┬───┐┌───┬───┐┌───┬───┐┌───┐
+ * │17 ││11 ││35 ││...││23 │ 1 ││29 │ 1 ││ 0 │ 2 ││ 0 │ 2 ││ 0 │ 1 ││...│
+ * └───┘└───┘└───┘└───┘└───┴───┘└───┴───┘└───┴───┘└───┴───┘└───┴───┘└───┘
+ *   │    │    │         ▲        ▲        ▲        ▲        ▲
+ *   └────┼────┼─────────┼────────┴────────┼────────┘        │
+ *        └────┼─────────┴─────────────────┘                 │
+ *             └─────────────────────────────────────────────┘
+ */
+export class SharedTypeMap<TItemType, THash, TAddress: number>
+  implements Iterable<TAddress> {
+  /**
+   * The header for the `SharedTypeMap` comprises 2 4-byte chunks:
+   *
+   * struct SharedTypeMapHeader {
+   *   int capacity;
+   *   int count;
+   * }
+   *
+   * ┌──────────┬───────┐
+   * │ CAPACITY │ COUNT │
+   * └──────────┴───────┘
+   */
+  static HEADER_SIZE: number = 2;
+  /** The offset from the header where the capacity is stored. */
+  static #CAPACITY: 0 = 0;
+  /** The offset from the header where the count is stored. */
+  static #COUNT: 1 = 1;
+
+  /**
+   * Each item in `SharedTypeMap` comprises 2 4-byte chunks:
+   *
+   * struct Node {
+   *   int next;
+   *   int type;
+   * }
+   *
+   * ┌──────┬──────┐
+   * │ NEXT │ TYPE │
+   * └──────┴──────┘
+   */
+  static ITEM_SIZE: number = 2;
+  /** The offset at which a link to the next item in the same bucket is stored. */
+  static #NEXT: 0 = 0;
+  /** The offset at which an item's type is stored. */
+  static #TYPE: 1 = 1;
+
+  /** The number of items to accommodate per hash bucket. */
+  static BUCKET_SIZE: number = 2;
+
+  data: Uint32Array;
```

Comment:
There would be significant memory saved if we used `DataView` instead, which would allow us to store the `type` field in 1 byte, but it also would complicate the implementation significantly.

The savings would come from the following:
- each node adds an entry for every unique edge type it is connected to, so in the worst case (every node had at least one edge of every type), that's `n * t * 3` bytes saved, where `n` is the number of nodes and `t` is the number of unique edge types in the graph.
- each edge keeps its type in its entry, which would be `e * 3` bytes saved, where `e` is the number of edges in the graph.

Additionally, typed arrays are _technically_ not portable, since they use the originating system's [endianess](https://en.wikipedia.org/wiki/Endianness)

Some of the reasons we haven't implemented this yet:
- `DataView` is likely not as well optimized as type arrays
- memory allocation is more complex, since we cannot rely on the uniform item size of a typed array
- In practice, most systems are LE, so the non-portable nature of typed array is probably a non-issue